src/server.c

```diff
@@ -5938,6 +5945,58 @@ void linuxMemoryWarnings(void) {
     }
 }
 
+void linuxTimeWarnings(void) {
+#ifdef HAVE_PROC_STAT
+#define PROC_STAT_STIME_IDX 15
+    long long sys_time_ticks, sys_time_ticks_start;
+    unsigned long systime_us, system_hz, test_time_us;
+    struct timespec ts;
+    unsigned long long start_us;
+    system_hz = sysconf(_SC_CLK_TCK);
+
+    if (!get_proc_stat_ll(PROC_STAT_STIME_IDX, &sys_time_ticks_start))
+        return;
+    if (clock_gettime(CLOCK_MONOTONIC, &ts) < 0) {
+        return;
+    }
+    start_us = (ts.tv_sec * 1000000 + ts.tv_nsec / 1000);
+
+    /* clock_gettime() busy loop of 5 times system hz (for a system_hz of 100 this is 50ms).
+     * If our clocksource is configured correctly (vdso) this will result in no system calls.
+     * If our clocksource is inefficient it'll waste most of the busy loop in the kernel. */
+    test_time_us = 5 * 1000000 / system_hz;
+    while (1) {
+        unsigned long long d;
+        if (clock_gettime(CLOCK_MONOTONIC, &ts) < 0)
+            return;
+        d = (ts.tv_sec * 1000000 + ts.tv_nsec / 1000) - start_us;
+        if (d >= test_time_us) break;
+    }
+    if (!get_proc_stat_ll(PROC_STAT_STIME_IDX, &sys_time_ticks))
+        return;
+    /* Calculate how much time we spent in kernel system calls */
+    systime_us = (1000000 * (sys_time_ticks - sys_time_ticks_start)) / system_hz;
+
+    /* If more than half the time was in system calls we probably have an inefficient clocksource - print a warning */
+    if (systime_us * 2 >= test_time_us) {
+        char avail[1024] = "";
+        char curr[1024] = "";
+        FILE *fp = fopen("/sys/devices/system/clocksource/clocksource0/available_clocksource","r");
+        if (fp) {
+            if (fgets(avail,sizeof(avail),fp)) avail[strcspn(avail, "\n")] = 0;
+            fclose(fp);
+        }
+        fp = fopen("/sys/devices/system/clocksource/clocksource0/current_clocksource","r");
+        if (fp) {
+            if (fgets(curr,sizeof(curr),fp)) curr[strcspn(curr, "\n")] = 0;
+            fclose(fp);
+        }
+        serverLog(LL_WARNING,"WARNING slow system clocksource detected. This can result in degraded redis performance. You should consider changing the system's clocksource. Current clocksource: %s, available clocksources: %s.",
+                  curr, avail);
```

Comment:
maybe we wanna make it easier on people (like we did for THP and others), and give them an example of the bash command they should run.

Reply:
The problem is that we can't be sure running a specific command would result in the desired result. Even if the clocksource is changed it might be changed to a non-safe one, so I'm not sure.
This is the current message:
`WARNING slow system clocksource detected. This can result in degraded performance. Consider changing the system's clocksource. Current clocksource: %s. Available clocksources: %s.`
I can change it to:
`WARNING slow system clocksource detected. This can result in degraded performance. Consider changing the system's clocksource. Current clocksource: %s. Available clocksources: %s. For example: run the command 'echo tsc > /sys/devices/system/clocksource/clocksource0/current_clocksource' as root. To permanently change the system's clocksource you'll need to set the 'clocksource=' kernel command line argument.`

Reply:
i think it's better.. the text still indicates that we're not confident this is the right thing.. but for the majority of the users seeing this, they'll be able to solve the problem without a round trip to google search. and arguably for many of them, they don't have the time or capacity to really understand the problem and risks anyway.